### DISCUSSION TITLE
difference between randomForestSRC::get.cindex() and pec::cindex

### DISCUSSION THREAD

**Discussion by Marina300** [kogalur/randomForestSRC#128] [Q&A]
Hi, 

what is the difference between get.cindex() and pec::cindex? They return different outputs and it looks like the are based on different formulas. Which ones? pec::cindex is based on Harrell’s  cindex and it is the proportion of concordant pairs divided by the total number of possible evaluation pairs.

What about the cindex computed with randomForestSRC::get.index() ???

Look at these examples: 
_ library("randomForestSRC")
 data(veteran)
 v.obj <- rfsrc(Surv(time, status) ~ ., data = veteran,   ntree = 100, block.size = 1)
 get.cindex(time = veteran$time, censoring = veteran$status, predicted = v.obj$predicted.oob)
**[1] 0.2923782**_

_pec:: cindex(v.obj,formula=Surv(time, status) ~ ., data = veteran)
        AppCindex Pairs Concordant
rfsrc      **35.7**       18637       6654_

Thanx in advance. 

Cheers, Marina

**Comment by ishwaran** [comment]
The `get.cindex` computes Harrell's concordance index and is described in the following vignette 

https://luminwin.github.io/randomForestSRC/articles/survival.html

But I don't get the discrepancy that you find.  I am using randomForestSRC_2.14.0 and pec_2021.10.11

```
> o <- rfsrc(Surv(time,status)~.,veteran)
> get.cindex(o$yvar[,1],o$yvar[,2],o$predicted.oob)
> [1] 0.2957707
```

This also agrees with the returned summary output
```
> o
                             Analysis: RSF
                              Family: surv
                      Splitting rule: logrank *random*
       Number of random split points: 10
                          (OOB) CRPS: 0.0628449
   (OOB) Requested performance error: 0.29577067

```

As for pec,

```
> pec:::cindex(list(rsf=o),formula=Surv(time,status)~.,veteran)
    AppCindex Pairs Concordant
rsf      30.6 18637       5702

```

This is very close.  Whatever discrepancy could perhaps be due to the way ties are handled.  Anyways, I explained exactly how we calculate C-index.  You can contact Thomas Gerds and ask him how he computes the value in `pec`

**Reply by Marina300** [reply]
Thank you for your reply. 
By saying "This also agrees with the returned summary output", do you mean that 
**get.cindex(o$yvar[,1],o$yvar[,2],o$predicted.oob)** returns the same value as **"(OOB) Requested performance error**?

Strangely I get different numbers on my computer. Have a look at my output below: 
_> o <- rfsrc(Surv(time,status)~.,veteran)
> get.cindex(o$yvar[,1],o$yvar[,2],o$predicted.oob)_
**[1] 0.2895511**

and 
 _pec:::cindex(list(rsf=o),formula=Surv(time,status)~.,veteran)
The c-index for right censored event times
Prediction models:
rsf 
rsf 
Right-censored response of a survival model
No.Observations: 137 
Pattern:
                Freq
 event          128 
 right.censored 9   
Censoring model for IPCW: marginal model (Kaplan-Meier for censoring distribution) 
No data splitting: either apparent or independent test sample performance
Estimated C-index in % at time=999 
    AppCindex Pairs Concordant
rsf        **31 18637**       5776
Warning message:
In summary.Cindex(x, print = TRUE, ...) :
  The C-index is not proper for t-year predictions. Blanche et al. (2018), Biostatistics, 20(2): 347--357.
Consider using time-dependent AUC instead: riskRegression::Score_

 0.2895511 and 31 18637% are not extremely close. It might be also the influence of random seeds. 
What do you think?

cheers, Marina
